### PR TITLE
Happily handle empty text messages

### PR DESCRIPTION
**Pull request by Palid**
Fix https://github.com/vector-im/element-web/issues/18873

Video preview:  https://user-images.githubusercontent.com/3636685/133771030-3f8dda7b-f005-4e54-88e3-a100f48439f4.mp4

---
Here's what your changelog entry will look like:

## ✨ Features
 * Happily handle empty text messages ([\#6825](https://github.com/matrix-org/matrix-react-sdk/pull/6825)). Fixes vector-im/element-web#18873. Contributed by [Palid](https://github.com/Palid).

Preview: https://6151a52d07a454b864aff27b--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Review by t3chguy — APPROVED** [review]

**Comment by Palid** [comment]
@t3chguy re-review when you can please and click merge if it's fine. Thanks!

**Review by t3chguy — CHANGES_REQUESTED** [review]
This approach has been rejected in the past. https://github.com/matrix-org/matrix-react-sdk/pull/4623#issuecomment-647450384

**Comment by Palid** [comment]
> This approach has been rejected in the past. [#4623 (comment)](https://github.com/matrix-org/matrix-react-sdk/pull/4623#issuecomment-647450384)

@t3chguy see below please:
This requires an explicit action from the user though in the modal window. 
We had a chat with @daniellekirkwood and @janogarcia internally to figure out a good-enough flow and seems that this is very much 80/20 where we require user's consent to get rid of the message, so they won't delete it accidentally, yet doesn't allow for the problem in the issue you linked to where you can just craft a scammy message.

**Comment by t3chguy** [comment]
Then the PR title should be updated, and reviews should be requested from design/product, it also doesn't account for the fact that a redaction means no further edits can be made, and the user is not made aware of this.

**Review by nadonomy — CHANGES_REQUESTED** [review]
I had a quick play around with this and re-traced the concerns and thinking from the previous, related PR (https://github.com/matrix-org/matrix-react-sdk/pull/4623).

I really like asking for confirmation from the user given the destructive and potentially unexpected action, however:

1. The copy in the modal isn't great. It generically refers to events, and advises on interactions the user hasn't hit via this path (editing room topics).
2. Users don't really read unexpected modals properly, so would be good to exercise more brevity in the content.
3. Given 2, at a glance the text input is baffling.

So, is it trivial to use a modal with more focused content for this instance? I'd be happy with this copy, edited at runtime to screenshot, taking inspiration from delete file confirmation modals in operating systems:

Screenshot 2021-09-17 at 13 16 27

**Comment by janogarcia** [comment]
@nadonomy Completely agree with your comments on the deletion modal copy not being even remotely intuitive/friendly.. That said, I'd consider this a separate issue which should be handled on its own.

I mean, @Palid is reusing the same modal that is shown when deleting a message via the floating actions. If we introduce a separate modal for the case when deleting a message via message editing, then that would result in a fragmented experience for what ultimately is the same action, deleting a message. Also, there's a maintenance cost to it. So, my suggestion would be to update the modal for both cases/scenarios, but treat it as a separate issue, to avoid blocking this PR. What do you think?

**Comment by novocaine** [comment]
Having different copy for different errors isn't a significant maintenance cost.

**Comment by janogarcia** [comment]
@novocaine It's not exactly that. It's a confirmation dialog that is shown before deleting a message. The one below is the dialog shown when attempting to delete a text message via the message floating actions. What I'm saying is that Nad's concerns should probably apply to both cases/modals (deleting a text message either via the floating actions or the inline message editor). That's why I'm suggesting to treat this as a separate issue, it feels inconsistent to ask for a reason to delete the text message depending on where you initiated the action from.

delete-message-modal

**Comment by nadonomy** [comment]
@janogarcia Ok. Here's the thinking I went through to arrive at a different conclusion. Slightly exhaustive but might help us align on what's best to do from here:

> @nadonomy Completely agree with your comments on the deletion modal copy not being even remotely intuitive/friendly. That said, I'd consider this a separate issue which should be handled on its own.

On the contrary— I don't think it's a separate issue. The proposed modal is only accessed via the specific user journey of editing _your own_ message to be blank. The actions are cleanly encapsulated.

> If we introduce a separate modal for the case when deleting a message via message editing, then that would result in a fragmented experience for what ultimately is the same action, deleting a message.

I did consider this, however the alternative of iterating on the generic modal is problematic, as it has multiple considerations which would definitely bloat scope:

1. It uses the term 'event', and can be accessed when removing more than just messages (e.g. room topic changes).
2. I am assuming the 'reason' input field is necessary for moderation. It makes sense for removing other peoples messages, but does it make sense for your own messages?
3. Based on trust & safety needs it's possible we'll sophisticate the [Remove] interaction into [Remove] & [Report].

> Also, there's a maintenance cost to it.

Ack, it is tiny though. 

> So, my suggestion would be to update the modal for both cases/scenarios,

I'm not sure if a generic modal would ever serve all cases well with the points above. We can't exercise brevity and clarity for deleting your own messages while also serving the general contexts for all events.

> but treat it as a separate issue, to avoid blocking this PR (lately we've been trying to keep PR/issues atomic for more predictable handling during sprints and release cycles, instead of inadvertently scope creeping them).

Understood!

> What do you think?

I think there's 3 clear paths, in order of my preference:

1. Merge as is. Accept that the UX is still sub par for clearing messages while editing, and anticipate needing to spend more time on this in future.
2. Improve this journey with a new modal. Separately, anticipate evolving the timeline interactions over time.
3. Improve the modal more generally.

Those would be my order of preference, but ultimately happy to defer the decision to yourself & @Palid given you're doing the work.

**Comment by Palid** [comment]
@nadonomy I think merging current one just as a fix would be great and then improving the general deletion&modal flow as of your suggestions is the proper way to go. If I understand correctly, the problem is that the deletion modal for your own messages is very subpar so far, and it definitely makes sense to change it into something better, but maybe not in this pull request. It was supposed to only fix a bug, the added delete modal is a tiny enhancement that should've been there from the start. I don't really consider maintenance cost a problem, too - it's more about streamlined experience that isn't weird (why do users even have a reason for their own messages deletion - exactly as you typed above).

I'll gladly make a new issue for the things you suggested here, so we can entirely overhaul user's journey with message deletion as an enhancement, a papercut fix, instead of feature creeping this one.
Sounds good?

**Comment by janogarcia** [comment]
> Here's the thinking I went through to arrive at a different conclusion. Slightly exhaustive but might help us align on what's best to do from here

Yeah, we basically agree on everything, but thank you for the detailed reply and context. It really helps in understanding the historical reasons for the existing stuff.

> I think there's 3 clear paths, in order of my preference:

Agree, last one would be definitely out of scope for this iteration. I’m happy with either 1 and 2, whichever works best for Engineering in the short and medium term. Either way we’ll need to tackle the suboptimal delete message/event modal experience on a subsequent issue.

> The proposed modal is only accessed via the specific user journey of editing your own message to be blank. […] I am assuming the 'reason' input field is necessary for moderation. It makes sense for removing other peoples messages, but does it make sense for your own messages? […] I'm not sure if a generic modal would ever serve all cases well with the points above. We can't exercise brevity and clarity for deleting your own messages while also serving the general contexts for all events.

That’s exactly the underlying issue that we’d need to tackle on a separate issue.

In both cases you’d be deleting your own message (either via editing your message to be blank or a delete contextual action). So, we should re-evaluate the current flow for deleting own content from the timeline.

File: src/components/views/rooms/MessageComposer.tsx
Comment (t3chguy) [inline src/components/views/rooms/MessageComposer.tsx:605]: not sure what this change is doing

**Review by t3chguy — APPROVED** [review]
LGTM code-wise otherwise

File: src/components/views/rooms/MessageComposer.tsx
Comment (Palid) [reply src/components/views/rooms/MessageComposer.tsx]: Fixed, ty for catching! Wrong merge probably.

**Review by nadonomy — APPROVED** [review]
@Palid Yup sgtm. Would be great if we can translate the thinking from this PR into a new issue so we don't lose it. Apologies this PR got backlogged in my queue.

I removed myself from it and added @janogarcia to review/approve to merge whenever you're both happy to!

**Comment by nadonomy** [comment]
...or until auto-merge does it automatically. 😀

**Pull request merged**
